lib/Backend/arm64/LegalizeMD.cpp

```diff
@@ -315,7 +315,41 @@ IR::Instr * LegalizeMD::LegalizeLoad(IR::Instr *instr, uint opndNum, LegalForms
 
 void LegalizeMD::LegalizeIndirOffset(IR::Instr * instr, IR::IndirOpnd * indirOpnd, LegalForms forms, bool fPostRegAlloc)
 {
-   int32 offset = indirOpnd->GetOffset();
+    // For LEA, we have special handling of indiropnds
+    auto correctSize = [](IR::Instr* instr, IR::IndirOpnd* indirOpnd)
+    {
+        // If the base and index operands on an LEA are different sizes, grow the smaller one,
+        // matching the larger size but keeping the correct signedness for the type
+        IR::RegOpnd* baseOpnd = indirOpnd->GetBaseOpnd();
+        IR::RegOpnd* indexOpnd = indirOpnd->GetIndexOpnd();
+        if (   baseOpnd != nullptr
+            && indexOpnd != nullptr
+            && baseOpnd->GetSize() != indexOpnd->GetSize()
+            )
+        {
+            if (baseOpnd->GetSize() < indexOpnd->GetSize())
+            {
+                IRType largerType = indexOpnd->GetType();
+                Assert(IRType_IsSignedInt(largerType) || IRType_IsUnsignedInt(largerType));
+                IRType sourceType = baseOpnd->GetType();
+                IRType targetType = IRType_IsSignedInt(sourceType) ? IRType_EnsureSigned(largerType) : IRType_EnsureUnsigned(largerType);
+                IR::Instr* movInstr = Lowerer::InsertMove(baseOpnd->UseWithNewType(targetType, instr->m_func), baseOpnd, instr, false);
+                indirOpnd->SetBaseOpnd(movInstr->GetDst()->AsRegOpnd());
+            }
+            else
+            {
+                Assert(indexOpnd->GetSize() < baseOpnd->GetSize());
+                IRType largerType = baseOpnd->GetType();
+                Assert(IRType_IsSignedInt(largerType) || IRType_IsUnsignedInt(largerType));
+                IRType sourceType = indexOpnd->GetType();
+                IRType targetType = IRType_IsSignedInt(sourceType) ? IRType_EnsureSigned(largerType) : IRType_EnsureUnsigned(largerType);
+                IR::Instr* movInstr = Lowerer::InsertMove(indexOpnd->UseWithNewType(targetType, instr->m_func), indexOpnd, instr, false);
```

Comment:
The downside of correcting the problem during legalization is the need to insert this extra move. If the size was correct at it's source, we could avoid the extra instruction in any case where we could be sure that the top 32 bits were empty (after any 32 operations were performed). That said, until we have more rigorous testing for arm64 this is probably the better approach.

Reply:
In any case, we want this sort of behavior for release builds, even if we change it to assert in debug builds - after a call to legalize, we want to be able to trust that the instructions we have are legal.